rust/connlib/tunnel/src/dns.rs

```diff
@@ -108,76 +90,129 @@ impl<T, V> ResolveStrategy<T, DnsQueryParams, V> {
+                let mut ips = self.ip_provider.get_n_ipv4(4);
+                ips.extend_from_slice(&self.ip_provider.get_n_ipv6(4));
+                ips
+            })
+            .clone();
+        for ip in &ips {
+            self.ips_to_fqdn.insert(*ip, fqdn.clone());
+        }
 
-pub(crate) fn create_local_answer<'a>(
-    ips: &HashSet<IpAddr>,
-    packet: IpPacket<'a>,
-) -> Option<IpPacket<'a>> {
-    let datagram = packet.unwrap_as_udp();
-    let message = as_dns(&datagram).unwrap();
-    let question = message.first_question().unwrap();
-    let qtype = question.qtype();
-
-    #[allow(clippy::wildcard_enum_match_arm)]
-    let resource = match qtype {
-        Rtype::A => RecordData::A(
-            ips.iter()
+        ips
+    }
+
+    // This function will panic if it's called with an invalid PTR question
+    fn get_records<N: ToName>(&mut self, question: &Question<N>) -> Vec<RecordData<DomainName>> {
+        match question.qtype() {
+            Rtype::A => self
+                .get_or_assign_ips(question.qname().to_name())
+                .iter()
                 .copied()
                 .filter_map(get_v4)
                 .map(domain::rdata::A::new)
-                .collect(),
-        ),
-        Rtype::AAAA => RecordData::Aaaa(
-            ips.iter()
+                .map(RecordData::A)
+                .collect_vec(),
+
+            Rtype::AAAA => self
+                .get_or_assign_ips(question.qname().to_name())
+                .iter()
                 .copied()
                 .filter_map(get_v6)
                 .map(domain::rdata::Aaaa::new)
-                .collect(),
-        ),
-        _ => unreachable!(),
-    };
+                .map(RecordData::Aaaa)
+                .collect_vec(),
+            Rtype::PTR => {
+                let ip = reverse_dns_addr(&question.qname().to_name::<Vec<_>>().to_string())
+                    .expect("ptr isn't a valid");
+                let Some(fqdn) = self.ips_to_fqdn.get(&ip) else {
+                    debug_assert!(false, "we expect this function to be called only with PTR records for resource ips");
+                    return Vec::new();
+                };
+
+                vec![RecordData::Ptr(domain::rdata::Ptr::new(fqdn.clone()))]
+            }
+            _ => Vec::new(),
+        }
+    }
+
+    fn is_resource(&self, question: &Question<impl ToName>) -> bool {
+        // TODO: we can shave off time from the lookup if we keep a hashset of ips
+        match question.qtype() {
+            Rtype::PTR => reverse_dns_addr(&question.qname().to_name::<Vec<_>>().to_string())
+                .is_some_and(|addr| self.fqdn_to_ips.values().flatten().contains(&addr)),
+            _ => get_description(&question.qname().to_name(), &self.dns_resources).is_some(),
+        }
+    }
+
+    // TODO: we can save a few allocations here still
+    pub(crate) fn handle<'a>(
+        &mut self,
+        dns_mapping: &bimap::BiMap<IpAddr, DnsServer>,
```

Comment:
Did you consider moving this into the state of the resolver too? I think that makes sense. Those are the servers the resolver forwards to if it can't handle a query.

Can wait for a follow-up though :)

Reply:
yeah, let's do it in a follow up PR